import sys
import peppy
from Bio import SeqIO

singularity: "docker://continuumio/miniconda3"
#configfile: "config/config.yaml"
pepfile: "config/pep.yaml"

SAMPLES = [x.sample_name for x in pep.samples]
SUBSAMPLE_TABLE = pep.subsample_table[0]
TES = [x.id.split("#")[0] for x in SeqIO.parse(config.get("TRANSPOSON_FASTA"), "fasta")]

MAIN_CONFIG = config.get("MAIN_CONFIG",'config/config.yaml')

subworkflow custom_genome:
    workdir:
        "../../gte21-custom-genome/"
    snakefile:
        "../../gte21-custom-genome/workflow/Snakefile"
    configfile:
        MAIN_CONFIG


rule all:
    input:
        expand('results/copies/{s}.tsv',s=SAMPLES),
        expand("results/snps/{s}-snps.vcf", s=SAMPLES),

rule samtools_index:
    input:
        "{file}.bam"
    output:
        "{file}.bam.bai"
    params:
        "" # optional params string
    wrapper:
        "0.70.0/bio/samtools/index"

rule bcftools_index:
    input:
        "{file}.vcf.gz"
    output:
        "{file}.vcf.gz.tbi"
    params:
        extra="-t" # optional params string
    wrapper:
        "0.72.0/bio/bcftools/index"

rule samtools_flagstat:
    input:
        "{file}.bam"
    output:
        "{file}.bam.flagstat"
    wrapper:
        "0.72.0/bio/samtools/flagstat"

include: "rules/trim.smk"
include: "rules/align-wgs.smk"
include: "rules/depth.smk"
include: "rules/snps.smk"
